pkg/kubelet/kubelet_pods.go

```diff
@@ -1143,74 +1145,118 @@ type PodKiller interface {
-	pk.podKillingLock.Lock()
-	defer pk.podKillingLock.Unlock()
-	_, ok := pk.mirrorPodTerminationMap[string(uid)]
-	return ok
+// IsPodPendingTerminationByUID checks whether the pod for the given uid is pending termination
+func (pk *podKillerWithChannel) IsPodPendingTerminationByUID(uid types.UID) bool {
+	pk.podKillingLock.RLock()
+	defer pk.podKillingLock.RUnlock()
+	if _, ok := pk.podTerminationMap[string(uid)+runningPodKey]; ok {
+		return true
+	}
+	if _, ok := pk.podTerminationMap[string(uid)+apiPodKey]; ok {
+		return true
+	}
+	return false
 }
 
-// IsMirrorPodPendingTerminationByPodName checks whether the given pod is in grace period of termination
-func (pk *podKillerWithChannel) IsMirrorPodPendingTerminationByPodName(podFullname string) bool {
-	pk.podKillingLock.Lock()
-	defer pk.podKillingLock.Unlock()
-	for _, name := range pk.mirrorPodTerminationMap {
+// IsPodPendingTerminationByPodName checks whether the given pod is in grace period of termination
+func (pk *podKillerWithChannel) IsPodPendingTerminationByPodName(podFullname string) bool {
+	pk.podKillingLock.RLock()
+	defer pk.podKillingLock.RUnlock()
+	for _, name := range pk.podTerminationMap {
 		if name == podFullname {
 			return true
 		}
 	}
 	return false
 }
 
-func (pk *podKillerWithChannel) markMirrorPodTerminated(uid string) {
+func (pk *podKillerWithChannel) markAPIPodTerminated(pod *v1.Pod) {
+	uid := string(pod.UID) + apiPodKey
+	klog.V(4).Infof("marking api pod termination %q", uid)
 	pk.podKillingLock.Lock()
-	klog.V(4).Infof("marking pod termination %q", uid)
-	delete(pk.mirrorPodTerminationMap, uid)
+	delete(pk.podTerminationMap, uid)
 	pk.podKillingLock.Unlock()
 }
 
-// MarkMirrorPodPendingTermination marks the pod entering grace period of termination
-func (pk *podKillerWithChannel) MarkMirrorPodPendingTermination(pod *v1.Pod) {
-	fullname := kubecontainer.GetPodFullName(pod)
-	klog.V(3).Infof("marking pod pending termination %q", string(pod.UID))
+func (pk *podKillerWithChannel) markRunningPodTerminated(pod *kubecontainer.Pod) {
+	uid := string(pod.ID) + runningPodKey
+	klog.V(4).Infof("marking running pod termination %q", uid)
 	pk.podKillingLock.Lock()
-	pk.mirrorPodTerminationMap[string(pod.UID)] = fullname
+	delete(pk.podTerminationMap, uid)
 	pk.podKillingLock.Unlock()
 }
 
+// CheckAndMarkPodPendingTerminationByPodPair checks to see if the pod is being
+// killed and returns true if it is, otherwise the pod is added to the map and
+// returns false
+func (pk *podKillerWithChannel) CheckAndMarkPodPendingTerminationByPodPair(podPair *kubecontainer.PodPair) bool {
+	pk.podKillingLock.Lock()
+	defer pk.podKillingLock.Unlock()
+	var apiPodExists bool
+	var runningPodExists bool
+	if podPair.APIPod != nil {
+		uid := string(podPair.APIPod.UID) + apiPodKey
+		_, apiPodExists = pk.podTerminationMap[uid]
+		if !apiPodExists {
+			klog.V(3).Infof("marking api pod pending termination %q", uid)
```

Comment:
Would the full name of the pod be relevant for the log message instead of the uid?

Reply:
We are tracking by UID here... I think we should have the UID, and I'll add the fullname as well.